### PR TITLE
feat(pa): dynamic backscraper and update to new source

### PR DESCRIPTION
**Pull request by grossir**
Implemented a new scraper targeting the API instead of the RSS feed. Since we needed to backscrape, that way we can target custom dates Also, updated the example files

Helps solve #967

### CODE REVIEW

**Review by quevon24 — COMMENTED** [review]
@grossir  @flooie I think it works great, but i have a question, in many results we have different opinions of the same case and i was wondering if it should be returned as a single item(case with multiple opinions) or separate items as it is now

for example, here: https://www.pacourts.us/api/opinion?startDate=2021-05-27T00%3A00%3A00-05%3A00&endDate=2021-07-15T00%3A00%3A00-05%3A00&courtType=SUPREME&boardDocketNumber=undefined&courtDocketNumber=undefined&keywords=undefined&postTypes=cd%2Cco%2Cdo%2Cmo%2Coaj&publicationType=undefined&sortDirection=-1

We have a majority opinion, concurring and dissent opinion and I think the management command in CL only creates one opinion, so i think those would be created as independent opinions instead of having them all in one cluster. Or how is it handled for these cases?

```

{
"Author": null,
"BoardDocketNumber": null,
"Caption": "Commonwealth v. Cosby, Jr., W., Aplt. - No. 39 MAP 2020",
"CourtDocketNumber": null,
"CourtType": 3,
"DispositionDate": "2021-06-30T00:00:00",
"Keywords": null,
"UserIdentifier": "M.D. Prothonotary",
"UploadDate": "0001-01-01T00:00:00",
"PostedToday": false,
"Postings": [
{
"Id": 82106,
"AuthorId": "Wecht, David N.",
"OpinionId": 74452,
"FileName": "J-100-2020mo - 104821740139246918.pdf",
"ProcessedDate": "2021-06-30T00:00:00",
"PostingTypeId": "mo",
"PublicationTypeId": null,
"RenderedDate": "2021-06-30T00:00:00",
"SortOrder": 0,
"FileVersion": 3,
"Author": {
"Id": 0,
"AuthorName": "Justice David Wecht",
"AuthorCode": "Wecht, David N.",
"Selectable": true,
"SortOrder": 1450
},
"PostType": {
"Id": 0,
"PostingTypeCode": "mo",
"PostingTypeId": "Majority Opinion",
"SortOrder": null
},
"PublicationType": null
},
{
"Id": 82107,
"AuthorId": "Dougherty, Kevin M.",
"OpinionId": 74452,
"FileName": "J-100-2020cdo - 104821740139246932.pdf",
"ProcessedDate": "2021-06-30T00:00:00",
"PostingTypeId": "cd",
"PublicationTypeId": null,
"RenderedDate": "2021-06-30T00:00:00",
"SortOrder": 0,
"FileVersion": 1,
"Author": {
"Id": 0,
"AuthorName": "Justice Kevin Dougherty",
"AuthorCode": "Dougherty, Kevin M.",
"Selectable": true,
"SortOrder": 1440
},
"PostType": {
"Id": 0,
"PostingTypeCode": "cd",
"PostingTypeId": "Concurring and Dissenting Opinion",
"SortOrder": null
},
"PublicationType": null
},
{
"Id": 82108,
"AuthorId": "Saylor, Thomas G.",
"OpinionId": 74452,
"FileName": "J-100-2020do - 104821740139246963.pdf",
"ProcessedDate": "2021-06-30T00:00:00",
"PostingTypeId": "do",
"PublicationTypeId": null,
"RenderedDate": "2021-06-30T00:00:00",
"SortOrder": 0,
"FileVersion": 1,
"Author": {
"Id": 0,
"AuthorName": "Justice Thomas G. Saylor",
"AuthorCode": "Saylor, Thomas G.",
"Selectable": false,
"SortOrder": 0
},
"PostType": {
"Id": 0,
"PostingTypeCode": "do",
"PostingTypeId": "Dissenting Opinion",
"SortOrder": null
},
"PublicationType": null
}
],
"CreatedById": null,
"DeletedById": null,
"UpdatedById": null,
"CreatedOn": null,
"DeletedOn": null,
"UpdatedOn": null,
"CreatedBy": null,
"DeletedBy": null,
"UpdatedBy": null,
"Id": 74452
},


```

**Comment by grossir** [comment]
This is indeed an enhancement we've been tracking for months. The OpinionCluster model has a one-to-many relation to the Opinion model, so it is what we should be doing. However, we would need to change both Juriscraper's OpinionSite(Linear) and cl_scrape_opinions for it to work. I bundled this and other changes into a proposed new scraper/site class some time ago, but is still pending of review. You can check it some more here https://github.com/freelawproject/juriscraper/issues/883#issuecomment-1981884866

**Comment by flooie** [comment]
@grossir can we figure out the conflicts here and does this need the enhanced v3 juriscraper?

**Comment by grossir** [comment]
To actually return and use OpinionClusters, we will indeed need a different approach both in juriscraper and courtlistener, and that's not ready yet

It seems there were no conflicts, I just merged main

File: juriscraper/opinions/united_states/state/pa.py [hunk 1]
```diff
@@ -5,74 +5,135 @@
...
+        start = now - timedelta(days=7)
+        self.params = {
+            "startDate": start.strftime(self.api_dt_format),
+            "endDate": now.strftime(self.api_dt_format),
+            "courtType": self.court.upper(),
+            "boardDocketNumber": "undefined",
+            "courtDocketNumber": "undefined",
+            "keywords": "undefined",
+            "postTypes": "cd,co,do,mo,oaj",
+            "publicationType": "undefined",
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/pa.py:41]: the undefined fields can simply be removed

File: juriscraper/opinions/united_states/state/pa.py [hunk 2]
```diff
@@ -5,74 +5,135 @@
...
+        now = datetime.now() + timedelta(days=1)
+        start = now - timedelta(days=7)
+        self.params = {
+            "startDate": start.strftime(self.api_dt_format),
+            "endDate": now.strftime(self.api_dt_format),
+            "courtType": self.court.upper(),
+            "boardDocketNumber": "undefined",
+            "courtDocketNumber": "undefined",
+            "keywords": "undefined",
+            "postTypes": "cd,co,do,mo,oaj",
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/pa.py:40]: we should expand the post Types to 
`cd,cds,co,cs,dedc,do,ds,mo,oaj,pco,rv,sd`

PA does a weird thing where they split apart lots of things

File: juriscraper/opinions/united_states/state/pa.py [hunk 3]
```diff
@@ -5,74 +5,135 @@
...
-            "[contains(title/text(), 'No.')]"
+        self.regex = re.compile(r"(.*)(?:[,-]?\s+Nos?\.)(.*)")
+        self.status = "Published"
+
+        now = datetime.now() + timedelta(days=1)
+        start = now - timedelta(days=7)
+        self.params = {
+            "startDate": start.strftime(self.api_dt_format),
+            "endDate": now.strftime(self.api_dt_format),
+            "courtType": self.court.upper(),
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/pa.py:36]: Upper doesnt appear to be needed ...

File: juriscraper/opinions/united_states/state/pa.py [hunk 4]
```diff
@@ -5,74 +5,135 @@
...
+
+        for cluster in json_response:
+            title = cluster["Caption"]
+            disposition_date = cluster["DispositionDate"].split("T")[0]
+            name, docket = self.parse_case_title(title)
+
+            for op in cluster["Postings"]:
+                author_str = op["Author"]["AuthorName"]
+                url = self.document_url.format(self.court, op["FileName"])
+                op_type = op["PostType"]["PostingTypeId"]
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/pa.py:64]: currently this is unused.

File: juriscraper/opinions/united_states/state/pasuperct.py [hunk 5]
```diff
@@ -7,11 +7,33 @@
...
 class Site(pa.Site):
+    court = "Superior"
+    days_interval = 20
+    first_opinion_date = datetime(1998, 2, 15)
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.court_id = self.__module__
-        self.url = "https://www.pacourts.us/Rss/Opinions/Superior/"
+        self.params["postTypes"] = "1,2,3,4,5,7,8,9,13,14,15,16,17,18"
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/pasuperct.py:25]: I think this should be 

1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,32,33

here they list all memorandum opinions as non-precedential - and dont use the opinion but I think that is what they are -  not Sure what complete and revised definitively mean

File: juriscraper/opinions/united_states/state/pacommwct.py [hunk 6]
```diff
@@ -14,17 +14,21 @@
...
         self.court_id = self.__module__
-        self.url = "https://www.pacourts.us/Rss/Opinions/Commonwealth/"
-        self.set_regex(r"(.*)(?:- |et al.\s+)(\d+.*\d{4})")
-        self.base = (
-            "//item[not(contains(title/text(), 'Judgment List'))]"
-            "[not(contains(title/text(), 'Reargument Table'))]"
-            "[not(contains(title/text(), 'Order Amending Rules'))]"
-        )
+        self.regex = re.compile(r"(.*)(?:- |et al.\s+)(\d+.*\d{4})")
+        self.params["postTypes"] = "complete,mo,opc,pn,sjo"
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/pacommwct.py:33]: I think this is right.

**Review by flooie — CHANGES_REQUESTED** [review]
Looks good but needs a few changes.

**Comment by grossir** [comment]
@flooie I updated the PR with the suggested changes; also added support for getting "per_curiam"

**Comment by flooie** [comment]
This is great thanks @grossir

**Review by flooie — APPROVED** [review]

**Pull request merged**